### PR TITLE
Draft: [pgx][postgres] Config.SchemaName refactoring usage

### PR DESCRIPTION
**Pull request by stephane-klein**
Follow-up this #547 issue, I try in this Merge Request, to refactor `Config.SchemaName` use in `pgx` and `postgres` modules.

The [`Config.SchemaName`](https://github.com/golang-migrate/migrate/blob/master/database/pgx/pgx.go#L48) is only used by [`Lock()`](https://github.com/golang-migrate/migrate/blob/3d5d577b61933c0b7a28cad41058d42dc0c87abb/database/pgx/pgx.go#L207) and [`Unlock()`](https://github.com/golang-migrate/migrate/blob/3d5d577b61933c0b7a28cad41058d42dc0c87abb/database/pgx/pgx.go#L207) functions (The `Config.SchemaName` was introduced in this [commit](https://github.com/golang-migrate/migrate/commit/16d63e3a760ede4e216f53105fce0df0f14037e8)).

In this Merge Request I want to compute `Config.SchemaName` and `Config.TableName` in `WithInstance` function, from this information:

- `Config.MigrationsTable`
- `Config.MigrationsTableQuoted`
- `SELECT CURRENT_SCHEMA()`

I suggest also to add `p.config.TableName` parameter to `database.GenerateAdvisoryLockId` because we can theoretically execute the `migrate.migration_table_1` and `migrate.migration_table_2` concurrently.

### CODE REVIEW

**Comment by stephane-klein** [comment]
Note: this Merge Request is in work in progress.

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:478]: 1. Use [raw strings](https://golang.org/ref/spec#String_literals) so we don't need to escape quotes.
2. The 2nd `p.config.SchemaName` should be `p.config.MigrationsTable`

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:49]: Clients shouldn't be able to set this so don't export this. Also, `tableName` isn't as descriptive so let's rename this to: `migrationsTableName`

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:48]: `SchemaName` can be set by clients so remove this comment

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:99]: We shouldn't set `config.SchemaName` using the schema found in `config.MigrationsTable` since 2 different schemas could be set by both `config.SchemaName` and `config.MigrationsTable`. e.g. `config.SchemaName = "my_schema1"` and `config.MigrationsTable = "\"my_schema2\".\"migrations_table\"`

We should probably add a `migrationsSchemaName` field to the `Config` struct and set that here. Its default value should be the current schema.

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for research and quick turn around for fixing this issue!

Once we settle on a fix, we'll need to port it to the postgres driver as well.

**Comment by stephane-klein** [comment]
> Thanks for research and quick turn around for fixing this issue!

Fixed.

> We'll need to port it to the postgres driver as well.

Done

---

@dhui need review 🙂

Note: please, don't merge directly this Pull Request, before merging I need to squash the commits + write a good commit message

Questions:

- I think that `Config.SchemaName` is confusing with the couple `Config.MigrationsTable` + `Config.MigrationsTableQuoted`, what do you think about to deprecate one of these methods ?
- There is no [URL Query parameter](https://github.com/golang-migrate/migrate/blob/4720914286210ebcabd01e8655e5fd8ea7565e73/database/postgres/README.md) to configure `Config.SchemaName`, then this parameter is only usable for users who embed *migrate* in their Golang project.
  Therefore, to be consistent, what do you think about choosing between:
     - add URL Query parameter to configure `Config.SchemaName`
     - or deprecate `Config.SchemaName` and detele it in future release ?

**Review by dhui — COMMENTED** [review]
> Note: please, don't merge directly this Pull Request, before merging I need to squash the commits + write a good commit message

👍 

> I think that Config.SchemaName is confusing with the couple Config.MigrationsTable + Config.MigrationsTableQuoted, what do you think about to deprecate one of these methods ?

I think we can keep all 3 config fields since they set different things and don't conflict with each other
* `SchemaName` sets the schema to be used when running migrations. This shouldn't affect `MigrationsTable`.
* `MigrationsTable` sets the table used to track the latest migration attempt
* `MigrationsTableQuoted` allows for greater flexibility when setting the migrations table e.g. allows setting the migrations table schema

> There is no URL Query parameter to configure Config.SchemaName, then this parameter is only usable for users who embed migrate in their Golang project.

The `search_path` url param allows you to set the schema so it's redundant.

File: database/pgx/pgx.go
Comment (stephane-klein) [reply database/pgx/pgx.go:108]: > this is unnecessary since it'll default to an empty string

Ok, I remove it (I included it because I like to follow [« Explicit is better than implicit »](https://www.python.org/dev/peps/pep-0020/) principle 😉)

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:228]: Note to self: document this backwards incompatible change
e.g. when updating to this version of `migrate`, you need to ensure that no migrations are currently running and no old versions of `migrate` will run at the same time as new versions.

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:118]: Actually, shouldn't this be moved out of this `else` block and before `if config.MigrationsTableQuoted`? I  think these should be the default values for `config.migrationsSchemaName` and `config.migrationsTableName` and only overwritten if `config.MigrationsTableQuoted` is specified. This will ensure that `config.migrationsSchemaName` is properly set.

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:479]: Should we use `p.quoteIdentifier()` here?
Also, I think we can remove the `if p.config.MigrationsTableQuoted` check from `quoteIdentifier()` since it's also used to quote all identifiers, not just the migrations table name.

**Review by dhui — CHANGES_REQUESTED** [review]
Sorry for the back and forth, but I think I found some other issues.

File: database/pgx/pgx.go
Comment (stephane-klein) [reply database/pgx/pgx.go:228]: > Note to self: document this backwards incompatible change

@dhui **suggestion:** what do you think to create un GitHub issue named « v4.15.0 release » and add this note in todo list in this issue?

File: database/pgx/pgx.go
Comment (stephane-klein) [reply database/pgx/pgx.go]: @dhui done 👍

File: database/pgx/pgx.go
Comment (stephane-klein) [reply database/pgx/pgx.go:478]: > Also, I think we can remove the if p.config.MigrationsTableQuoted check from quoteIdentifier() since it's also used to quote all identifiers, not just the migrations table name.

@dhui Yes if I replace `p.quoteIdentifier(p.config.MigrationsTable)` by 

```
...` + quoteIdentifier(p.config.migrationsSchemaName) + `.` + quoteIdentifier(p.config.migrationsTableName) + `...
```

everywhere.

I'm going to apply this change 👍

Comment (stephane-klein) [reply database/pgx/pgx.go]: Done

**Comment by stephane-klein** [comment]
@dhui need review 🙂

Do not merge, I need to squash the commit messages when you will validate the Pull Request.

File: database/pgx/pgx.go
Comment (dhui) [reply database/pgx/pgx.go:227]: No worries, I've create a draft release and add this note there

**Review by dhui — APPROVED** [review]
LGTM, thanks for the quick fix!
I'll merge after you've squashed.

**Comment by stephane-klein** [comment]
> MR https://gitlab.spacefill.fr/spacefill/spacefill/-/merge_requests/3746

@dhui  Done, thanks

**Review by dhui — APPROVED** [review]

**Pull request merged**
